experimental/puppeteer-firefox/lib/Browser.js

```diff
@@ -114,15 +114,21 @@ class Browser extends EventEmitter {
   }
 
   /**
-   * @param {function(!Target):boolean} predicate
+   * @param {function(!Target):boolean|Promise<boolean>} predicate
    * @param {{timeout?: number}=} options
    * @return {!Promise<!Target>}
    */
   async waitForTarget(predicate, options = {}) {
     const {
       timeout = 30000
     } = options;
-    const existingTarget = this.targets().find(predicate);
+    let existingTarget;
+    for (const target of this.targets()) {
+      if (await predicate(target)) {
```

Comment:
Could we also ensure that if predicate takes longer that the timeout, we abort the waitForTarget call?

Reply:
good advice